[inductor][non determinism] Disable autotuning when determinisitic mode is ON

anijain2305 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #99851
* #99793


This removes a source of non-determinism seen in `sebotnet33ts_256`. However, we should see if we can reduce non-determinism in autotuning in general.


cc @soumith @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

anijain2305 committed:
[inductor][non determinism] Clip benchmarking result of small kernels

[ghstack-poisoned]

anijain2305 committed:
Update on "[inductor][non determinism] Clip benchmarking result of small kernels"



This removes a source of non-determinism seen in `sebotnet33ts_256`. 


cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

anijain2305 commented:
Inductor run here - https://github.com/pytorch/pytorch/actions/runs/4781747485

anijain2305 commented:
@pytorchbot merge

anijain2305 committed:
Update on "[inductor][non determinism] Clip benchmarking result of small kernels"



This removes a source of non-determinism seen in `sebotnet33ts_256`. This is definitely not bulletproof. Large kernels w. very similar perf could also lead to non-determinism (we could use `round` instead of `clip` for that). 


cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

desertfire commented:
Should we introduce a flag to pick a deterministic triton-generated code (e.g. based on a sorting of hashcode) and enable that for accuracy testing?

ngimel commented:
We can just disable autotuning (there's a flag for that), that should be deterministic

anijain2305 committed:
Update on "[inductor][non determinism] Clip benchmarking result of small kernels"



This removes a source of non-determinism seen in `sebotnet33ts_256`. This is definitely not bulletproof. Large kernels w. very similar perf could also lead to non-determinism (we could use `round` instead of `clip` for that). 


cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

anijain2305 commented on torch/_inductor/triton_heuristics.py:
```diff
@@ -810,11 +819,13 @@ def persistent_reduction(size_hints, reduction_hint=False, meta=None, filename=N
         c.kwargs.pop("RBLOCK")
 
+    if not config.triton.autotune_pointwise:
+        configs = configs[:1]
```
Oh, this returns only 1 config if autotune_pointwise if False. Basically disables autotuning.

jansel commented on torch/_inductor/triton_heuristics.py:
```diff
@@ -168,6 +168,15 @@ def launcher({', '.join(def_args)}, grid, stream):
+        autotuning nondeterministic.  This function rounds the benched timing to
+        a default value of one decimal place, i.e., precision of 0.1 ms.
+        """
+        return builtins.round(x, decimal_places)
```
Rounding could still be non-determinstic (just less likely).  Perhaps we should change the cache format to just store which one was selected?

anijain2305 committed:
Update on "[inductor][non determinism] Round benchmarking autotuning results"



This removes a source of non-determinism seen in `sebotnet33ts_256`. We round the benchmarking numbers to 1 decimal place, precision of 0.1 ms.


cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

anijain2305 committed:
Update on "[inductor][non determinism] Round benchmarking autotuning results"



This removes a source of non-determinism seen in `sebotnet33ts_256`. We round the benchmarking numbers to 1 decimal place, precision of 0.1 ms.


cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

anijain2305 committed:
Update on "[inductor][non determinism] Disable autotuning when determinisitic mode is ON"



This removes a source of non-determinism seen in `sebotnet33ts_256`. However, we should see if we can reduce non-determinism in autotuning in general.


cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

desertfire reviewed (approved):
The new version LGTM.

anijain2305 commented:
@pytorchbot merge
